compiler/rustc_middle/src/ty/context.rs

```diff
@@ -423,6 +424,27 @@ pub struct TypeckResults<'tcx> {
     /// see `MinCaptureInformationMap` for more details.
     pub closure_min_captures: ty::MinCaptureInformationMap<'tcx>,
 
+    /// Tracks the fake reads required for a closure and the reason for the fake read.
+    /// When performing pattern matching for closures, there are times we don't end up
+    /// reading places that are mentioned in a closure (because of _ patterns). However,
+    /// to ensure the places are initialized, we introduce fake reads.
+    /// Consider these two examples:
+    /// ``` (discriminant matching with only wildcard arm)
+    /// let c = || match x { _ => ()};
+    /// ```
+    /// ``` (destructured assignments)
+    /// let c = || {
+    ///     let (t1, t2) = t;
+    /// }
+    /// ```
+    /// In the first example, we don't want to read/borrow `x` in `c` as the only match
+    /// arm contains only a wildcard.
```

Comment:
I think it might be useful to extend the first example to be:

```rust
let x: u8;
let c = || match x { _ => () };
```

then you can say "in this example, we don't need to actually read/borrow `x` in `c`, and so we don't want to capture it. However, we do still want an error here, because `x` should have to be initialized at the point where `c` is created. Therefore, we add a "fake read" instead."

Actually, I know we have tests for cases where `x` has type `!`, but do we have a test for a case like this? I imagine we want that.